### PR TITLE
Fix Possible Congestion Scenario in `SortPreservingMergeExec`

### LINKED ISSUE #12300: Congestion Scenario in SPM
### Is your feature request related to a problem or challenge?

During the initialization of cursors of SPM (SortPreservingMerge), it needs to get `Poll::Ready()` for each partition, to initiate the next partition cursor. 
https://github.com/apache/datafusion/blob/8db30e25d6fe65a9779d237cf48aea9aee297502/datafusion/physical-plan/src/sorts/merge.rs#L161-L166
We have experienced a scenario in our downstream where SPM is polling 1st partition continuously but never get a `Poll::Ready()`. Those polls get all upstream operator buffers and channels grow. However, if SPM have skipped that partition and tried to initiate other partitions before all those retrials, we did not experience this problem (congestion in partition 1 is cleared once the other partitions are pulled).

### Describe the solution you'd like

SPM does not wait initiating the same cursor partition. Instead, it iterates over all partitions continuously whether it is OK or not, until they are all successfully initiated.

### Describe alternatives you've considered

_No response_

### Additional context

I have a test which actually needs to pass:
https://github.com/synnada-ai/datafusion-upstream/blob/eb068a07020c44f5212e087269390652bd47e397/datafusion/core/tests/fuzz_cases/merge_fuzz.rs#L333

It simply tries to collect results of a plan which consist of an SPM and a mock source exec. That mock source returns pending from the 1st partition until the 2nd partition is polled. Once the 2nd is polled, 1st one also starts to send its results. In the current version of `poll_next_inner()` of `SortPreservingMergeStream`, once the `maybe_poll_stream()` returns s a pending for the first partition, it never enters `poll_next_inner()` again.

### PR DESCRIPTION
**Pull request by berkaysynnada**
## Which issue does this PR close?

Closes #12300.

## Rationale for this change

Please see the issue.

### CODE REVIEW

File: datafusion/physical-plan/src/sorts/merge.rs [hunk 1]
```diff
@@ -156,12 +164,22 @@ impl<C: CursorValues> SortPreservingMergeStream<C> {
...
+            let remaining_partitions = self.uninitiated_partitions.clone();
+            for i in remaining_partitions {
+                match self.maybe_poll_stream(cx, i) {
+                    Poll::Ready(Err(e)) => {
+                        self.aborted = true;
+                        return Poll::Ready(Some(Err(e)));
+                    }
+                    Poll::Pending => {
+                        self.uninitiated_partitions.rotate_left(1);
+                        cx.waker().wake_by_ref();
```
Comment (berkaysynnada) [inline datafusion/physical-plan/src/sorts/merge.rs:177]: I am not sure if this usage has some side-effects or decrease performance, but I cannot wake the SPM poll again once it receives a pending from its first partition

**Comment by berkaysynnada** [comment]
@alamb, you might be more familiar with these parts of the code. Do you have any ideas about this solution, or perhaps you could suggest a different approach?

File: datafusion/physical-plan/src/sorts/merge.rs [hunk 2]
```diff
@@ -97,6 +100,10 @@ pub(crate) struct SortPreservingMergeStream<C: CursorValues> {
 
     /// number of rows produced
     produced: usize,
+
+    /// Unitiated partitions. They are stored in a vector to keep them in
```
Comment (alamb) [inline datafusion/physical-plan/src/sorts/merge.rs:104]: Can you please document what an "uninitiated partition" means in this context? I think it means  partitions whose streams that have been polled haven't been ready yet

File: datafusion/physical-plan/src/sorts/merge.rs [hunk 3]
```diff
@@ -156,12 +164,22 @@ impl<C: CursorValues> SortPreservingMergeStream<C> {
         }
         // try to initialize the loser tree
         if self.loser_tree.is_empty() {
-            // Ensure all non-exhausted streams have a cursor from which
-            // rows can be pulled
-            for i in 0..self.streams.partitions() {
-                if let Err(e) = ready!(self.maybe_poll_stream(cx, i)) {
```
Comment (alamb) [inline datafusion/physical-plan/src/sorts/merge.rs:162]: It looks to me like the issue is that the `ready!` macro `return`s if the poll isn't ready -- and thus the other streams aren't polled. 

I wonder if simply changing this to not use `ready!` would work, and be a smaller change?

Like could you change this to be the following instead? I think that would ensure each stream is polled

```rust
                if let Poll::Ready(Err(e)) = ready!(self.maybe_poll_stream(cx, i)) {

File: datafusion/physical-plan/src/sorts/merge.rs [hunk 4]
```diff
@@ -156,12 +164,22 @@ impl<C: CursorValues> SortPreservingMergeStream<C> {
...
-                    self.aborted = true;
-                    return Poll::Ready(Some(Err(e)));
+            // Ensure all non-exhausted streams have a cursor from which rows can be pulled
```
Comment (alamb) [inline datafusion/physical-plan/src/sorts/merge.rs:167]: This comment implies to me that the code would / should `poll` *all* the streams. However, the code seems to ensure now that only streams that had previously not returned `Ready` for a poll are now polled.

**Review by alamb — COMMENTED** [review]
Thanks @berkaysynnada  -- it seems to me like the intent of the current code is to poll all the streams but after this change, the streams are only polled until they first return `Ready`

I left a suggestion on a way that maybe is closer to what the orignal intent was

File: datafusion/physical-plan/src/sorts/merge.rs (hunk 3 above)
Comment (berkaysynnada) [reply datafusion/physical-plan/src/sorts/merge.rs:162]: I guess you meant
`if let Poll::Ready(Err(e)) = self.maybe_poll_stream(cx, i)`
This change solves my problem, but some existing tests fails:
```
thread 'sorts::sort_preserving_merge::tests::test_async' panicked at datafusion/physical-plan/src/sorts/sort_preserving_merge.rs:959:9:
assertion `left == right` failed
  left: 0
 right: 1
```

```
assertion failed: poll.is_pending()
thread 'sorts::sort_preserving_merge::tests::test_drop_cancel' panicked at datafusion/physical-plan/src/test.rs:42:5:
assertion failed: poll.is_pending()
```

I suppose that's because of we do not return pending in case of any of the partitions returns pending (what I understand is if there is at least one pending partition, SPM should return pending).

File: datafusion/physical-plan/src/sorts/merge.rs (hunk 4 above)
Comment (berkaysynnada) [reply datafusion/physical-plan/src/sorts/merge.rs:167]: IMO, the behavior is more correct now. In the previous version, let's assume the 1st partition is exhausted and returns `None` without setting its cursor. Then, the 2nd partition returns `Pending`. When `poll_next_inner()` is polled again, the iteration starts from the 1st partition, which has already returned None AFAIK polling exhausted streams could cause problems). Therefore, I track which streams have returned a result (either `None` or `Some()`), and which ones have returned `Pending` only.

**Comment by tustvold** [comment]
I'm afraid I don't have capacity to review this, and am not likely to in the foreseeable future, however, one thing to be aware of is that SortPreservingMerge must be stable. Therefore if the first stream is not ready it must wait for it to be so, before it can proceed

File: datafusion/physical-plan/src/sorts/merge.rs (hunk 3 above)
Comment (alamb) [reply datafusion/physical-plan/src/sorts/merge.rs:162]: I coded up what I had in mind here: https://github.com/synnada-ai/datafusion-upstream/pull/34

File: datafusion/physical-plan/src/sorts/merge.rs (hunk 1 above)
Comment (alamb) [reply datafusion/physical-plan/src/sorts/merge.rs:179]: I did some research -- see https://github.com/synnada-ai/datafusion-upstream/pull/34/files#r1743621057

I think calling `wake_by_ref` effectively tells tokio to schedule this poll loop again after handling other tasks, which makes sense to me (as I am not sure how else we would signal to tokio that the merge is ready to go)

But I share your concern that this will cause some sort of performance issue

File: datafusion/physical-plan/src/sorts/merge.rs (hunk 4 above)
Comment (alamb) [reply datafusion/physical-plan/src/sorts/merge.rs:169]: see https://github.com/synnada-ai/datafusion-upstream/pull/34 for alternate idea

Comment (berkaysynnada) [reply datafusion/physical-plan/src/sorts/merge.rs:169]: I've tried to explain my concern with that: https://github.com/synnada-ai/datafusion-upstream/pull/34#issuecomment-2330825156

Comment (alamb) [reply datafusion/physical-plan/src/sorts/merge.rs:169]: 👍  -- response in https://github.com/synnada-ai/datafusion-upstream/pull/34#issuecomment-2331231064

File: datafusion/core/tests/fuzz_cases/merge_fuzz.rs [hunk 5]
```diff
@@ -160,3 +179,142 @@ fn concat(mut v1: Vec<RecordBatch>, v2: Vec<RecordBatch>) -> Vec<RecordBatch> {
...
+}
+
+impl RecordBatchStream for CongestedStream {
+    fn schema(&self) -> SchemaRef {
+        Arc::clone(&self.schema)
+    }
+}
+
+#[tokio::test]
+async fn test_spm_congestion() -> Result<()> {
```
Comment (alamb) [inline datafusion/core/tests/fuzz_cases/merge_fuzz.rs:292]: I read this test a bit more -- it doesn't seem like it is actually a fuzz test (aka it doesn't seem to have any random inputs, for example). 

I think it would make more sense to put it with the other sort preserving merge tests: https://github.com/apache/datafusion/blob/6034be42808b43e3f48f6e58ec38cc35fa253abb/datafusion/physical-plan/src/sorts/sort_preserving_merge.rs#L301-L302

**Review by alamb — APPROVED** [review]
Per the discussion on https://github.com/synnada-ai/datafusion-upstream/pull/34 I am now convinced that this design is reasonable and an improvement over the current one.

In my opinion it is important for this PR:
1. Encode the rationale for polling streams
2. Fix the comments about "ensure all non-exhausted streams" to the new intent (something like "ensure all streams have been polled at least once to start processing") or something

It would be nice to have to:
1. Move the tests with the existing sort preserving merge tests (rather than fuzz)
2. Add a new test that verifies the "doesn't poll after `Ready(None)` is returned" as that seems an important property that is not currently covered


Thank you for bearing with me @berkaysynnada

**Comment by berkaysynnada** [comment]
I’ve incorporated the final feedback. If there aren’t any more suggestions, I think we are good to merge this PR once the tests pass.

**Review by alamb — APPROVED** [review]
I think the PR looks (really) good to me now -- thank you @berkaysynnada
